packages/postgres-database/src/simcore_postgres_database/utils_projects_nodes.py

```diff
@@ -0,0 +1,215 @@
+import uuid
+from dataclasses import asdict, dataclass, field
+
+import sqlalchemy
+from aiopg.sa.connection import SAConnection
+from sqlalchemy import literal_column
+
+from .errors import ForeignKeyViolation, UniqueViolation
+from .models.projects_nodes import projects_nodes
+from .models.projects_to_projects_nodes import projects_to_projects_nodes
+
+
+#
+# Errors
+#
+class BaseProjectsNodesError(Exception):
+    ...
+
+
+class ProjectsNodesProjectNotFound(BaseProjectsNodesError):
+    ...
+
+
+class ProjectsNodesNodeNotFound(BaseProjectsNodesError):
+    ...
+
+
+class ProjectsNodesOperationNotAllowed(BaseProjectsNodesError):
+    ...
+
+
+class ProjectsNodesDuplicateNode(BaseProjectsNodesError):
+    ...
+
+
+@dataclass(frozen=True, slots=True, kw_only=True)
+class ProjectsNodeCreate:
+    node_id: uuid.UUID
+    required_resources: dict = field(default_factory=dict)
+
+
+@dataclass(frozen=True, slots=True, kw_only=True)
+class ProjectsNode(ProjectsNodeCreate):
+    created: datetime.datetime
+    modified: datetime.datetime
+
+
+@dataclass(frozen=True, slots=True, kw_only=True)
+class ProjectsNodesRepo:
+    project_uuid: uuid.UUID
+
+    async def create(
+        self, connection: SAConnection, *, node: ProjectsNodeCreate
+    ) -> ProjectsNode:
+        """creates a new entry in *projects_noeds* and *projects_to_projects_nodes* tables
+
+        NOTE: Do not use this in an asyncio.gather call as this will fail!
+
+        Raises:
+            ProjectsNodesProjectNotFound: in case the project_uuid does not exist
+            ProjectsNodesDuplicateNode: in case the node already exists
+
+        """
+        async with connection.begin():
+            try:
+                result = await connection.execute(
+                    projects_nodes.insert()
+                    .values(**asdict(node))
+                    .returning(literal_column("*"))
+                )
+                created_node_db = await result.first()
+                assert created_node_db  # nosec
+                created_node = ProjectsNode(**dict(created_node_db.items()))
+
+                result = await connection.execute(
+                    projects_to_projects_nodes.insert().values(
+                        project_uuid=f"{self.project_uuid}",
+                        node_id=f"{created_node.node_id}",
+                    )
+                )
+                assert result.rowcount == 1  # nosec
+                return created_node
+            except ForeignKeyViolation as exc:
+                # this happens when the project does not exist
+                raise ProjectsNodesProjectNotFound(
+                    f"Project {self.project_uuid} not found"
+                ) from exc
+            except UniqueViolation as exc:
+                # this happens if the node already exists
+                raise ProjectsNodesDuplicateNode(
+                    f"Project node {node.node_id} already exists"
+                ) from exc
+
+    async def add(
```

Comment:
THOUGHT: I have in mind the CRUD operations here and find it difficult (without looking at the code) to distinguish 
 between `create` and `add` provided that the repo is called `ProjectsNodesRepo` (i.e. with projects in plural!):
- should it rather be `ProjectNodesRepo` (i.e. projects in singular) since this repo only handles one single project at a time?
- how can I `add` a node the project without having it created it first in the `project_nodes` table? Actually this member function could be even used inside ``create`` to produce the mapping , right?

Reply:
yes, I've been thinking of not having create at all actually and only add. also remove instead of delete. 
The repo talks to the table *projects_nodes* such that the listing of the tables is not having singular/plural confusion/mixup.
add is used to attach a node from another project, i.e. for sharing nodes among projects.

Reply:
corrected, and using add instead of create